peer_test.go

```diff
@@ -0,0 +1,450 @@
+		voteGranted:     false,
+		matchIndex:      10,
+		verifiedCounter: 90,
+	}
+	_, err := oneRPC(tp, nil, &reply, expProgress)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestPeer_RequestVoteRPC_newTerm(t *testing.T) {
+	tp := makePeerTesting(t, &TestingPeer{
+		initControl:  &requestVoteControl,
+		initProgress: &requestVoteProgress,
+	})
+	reply := RequestVoteResponse{
+		Term:    13,
+		Granted: true,
+	}
+	expProgress := peerProgress{
+		peerID:          tp.peerID,
+		term:            13,
+		voteGranted:     false,
+		matchIndex:      0,
+		verifiedCounter: 90,
+	}
+	_, err := oneRPC(tp, nil, &reply, expProgress)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestPeer_RequestVoteRPC_confirmError(t *testing.T) {
+	control := requestVoteControl
+	tp := makePeerTesting(t, &TestingPeer{
+		initControl:  &control,
+		initProgress: &requestVoteProgress,
+	})
+	rpc := makeRequestVoteRPC(tp.peer)
+	err := rpc.prepare(tp.peer.shared, tp.peer.control)
+	if err != nil {
+		t.Fatalf("Unexpected error in prepare: %v", err)
+	}
+
+	control.role = Follower
+	tp.controlCh <- control
+	tp.peer.drainNonblocking()
+	err = rpc.confirm(tp.peer)
+	if err == nil || !strings.Contains(err.Error(), "no longer candidate") {
+		t.Fatalf("Expected cancel due to not candidate, got %v", err)
+	}
+
+	control.role = Candidate
+	control.term++
+	tp.controlCh <- control
+	tp.peer.drainNonblocking()
+	err = rpc.confirm(tp.peer)
+	if err == nil || !strings.Contains(err.Error(), "term changed") {
+		t.Fatalf("Expected cancel due to larger term, got %v", err)
+	}
+}
+
+func TestPeer_RequestVoteRPC_sendRecvError(t *testing.T) {
+	tp := makePeerTesting(t, &TestingPeer{
+		initControl:  &requestVoteControl,
+		initProgress: &requestVoteProgress,
+	})
+	err := oneErrRPC(tp, nil, fmt.Errorf("a transport error"))
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+///////////////////////// AppendEntries /////////////////////////
+
+var appendEntriesControl = peerControl{
+	term:          12,
+	role:          Leader,
+	lastIndex:     3,
+	lastTerm:      9,
+	verifyCounter: 120,
+}
+
+var appendEntriesProgress = peerProgress{
+	term:            12,
+	matchIndex:      1,
+	verifiedCounter: 90,
+}
+
+func TestPeer_AppendEntries_heartbeat(t *testing.T) {
+	tp := makePeerTesting(t, &TestingPeer{
+		initControl:  &appendEntriesControl,
+		initProgress: &appendEntriesProgress,
+	})
+	exp := AppendEntriesRequest{
+		Term:         12,
+		Leader:       tp.localTrans.EncodePeer(tp.localAddr),
+		PrevLogEntry: 3,
+		PrevLogTerm:  9,
+	}
+	reply := AppendEntriesResponse{
+		Term:    12,
+		LastLog: 0,
+		Success: true,
+	}
+	expProgress := peerProgress{
+		peerID:          tp.peerID,
+		term:            12,
+		matchIndex:      3,
+		verifiedCounter: 120,
+	}
+	_, err := oneRPC(tp, &exp, &reply, expProgress)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TODO: more tests for AppendEntries
+
+///////////////////////// InstallSnapshot /////////////////////////
+
+// TODO: more tests for InstallSnapshot
```

Comment:
I don't know of any blockers to writing these, but I think it makes sense to get a review on the basic approach first.